features/steps/helpers.py

```diff
@@ -23,27 +27,40 @@ def sign_in(context, login=None, browser=None):
     submit_button.click()
 
 
-def cautious_click(browser, xpath, iterations=10, timeout=1):
-    iterations = iterations - 1
-    element = browser.find_element_by_xpath(xpath)
-    try:
-        sleep(timeout)
-        element.click()
-    except ElementClickInterceptedException as Error:
-        if iterations > 0:
-            cautious_click(browser, xpath, iterations, timeout)
-        else:
-            raise Error
-
-
-def cautious_send_keys(browser, xpath, sequence, iterations=10, timeout=1):
-    iterations = iterations - 1
-    element = browser.find_element_by_xpath(xpath)
-    try:
-        sleep(timeout)
-        element.send_keys(sequence)
-    except ElementNotInteractableException as Error:
-        if iterations > 0:
-            cautious_send_keys(browser, xpath, sequence, iterations, timeout)
-        else:
-            raise Error
+def cautious_click(browser, xpath, iterations=CAUTIOUS_ACTION_ITERATIONS, timeout=CAUTIOUS_ACTION_TIMEOUT):
+    cautious_action(browser, xpath, "click", [], ElementClickInterceptedException, iterations, timeout)
+
+
+def cautious_send_keys(
+    browser, xpath, sequence, iterations=CAUTIOUS_ACTION_ITERATIONS, timeout=CAUTIOUS_ACTION_TIMEOUT
+):
+    cautious_action(browser, xpath, "send_keys", [sequence], ElementNotInteractableException, iterations, timeout)
+
+
+def cautious_action(
```

Comment:
it can be simplified a bit - it is just a suggestion ;)
```python
def cautious_action(
    browser,
    xpath,
    action,
    expected_exception,
    iterations,
    timeout,
    *args,
    **kwargs,
):
    received_exception = None
    while iterations > 0:
        try:
            element = browser.find_element_by_xpath(xpath)
            sleep(timeout)
            getattr(element, action)(*args, **kwargs)
            break
        except expected_exception as error:
            received_exception = error
            iterations -= 1
    else:
        raise received_exception
```
please not, that we can decide to use any of `*args`, `**kwargs`

Reply:
Thanks!
Wasn't aware that `*args, **kwargs` solution would work here. Thought it's some higher level magic involving Django modules.
Never heard of **while-else** before! 👀 What a peculiar mechanism!
Also, at some point I was advised to avoid breaks in general. I was told they can make it harder on following the code and that it's better to just write proper condition for the loop. What's your opinion on this?